### DISCUSSION TITLE
dsl::integer and negative numbers

### DISCUSSION THREAD

**Discussion by facontidavide** [foonathan/lexy#87] [Q&A] [answered]
Hi,

I notice that `dsl::integer` is not parsing negative numbers. Is that the desired behavior?

I am not sure how to address this. This is my sample code:

```c++
#include <lexy/action/parse.hpp>
#include <lexy/callback.hpp>
#include <lexy/dsl.hpp>
#include <lexy_ext/report_error.hpp>
#include <lexy/input/string_input.hpp>

namespace dsl = lexy::dsl;

struct production
{
  static constexpr auto whitespace = dsl::ascii::space;

  static constexpr auto rule =
      [] {
        auto integer = dsl::integer<int64_t>;
        auto hex_integer = LEXY_LIT("0x") >> dsl::integer<int, dsl::hex>;

        return (hex_integer | integer) + dsl::eof;
      }();
  static constexpr auto value = lexy::forward<int>;
};

int main()
{
  const char* strings[] = {"7", "0x78", "-7"};
  for( const auto& str: strings )
  {
    auto input  = lexy::zstring_input(str);
    auto result = lexy::parse<production>(input, lexy_ext::report_error);
    if (result.has_value())
    {
      std::printf("%d\n", result.value());
    }
  }
  return 0;
}
```

Output:

```
7
120
error: while parsing Integer
     |
   1 | -7
     | ^ exhausted choice
```

**Comment by facontidavide** [comment, marked as answer]
I found the solution looking at the **json** example, but I am not quite sure why the `dsl::peek` is required.

```c++
struct production
{
  struct integer
  {
    static constexpr auto rule = dsl::sign + dsl::integer<int64_t>;
    static constexpr auto value = lexy::as_integer<int64_t>;
  };

  static constexpr auto rule = [] {
    auto hex_integer = LEXY_LIT("0x") >> dsl::integer<int64_t, dsl::hex>;
    auto regular_integer = dsl::peek(dsl::lit_c<'-'> / dsl::lit_c<'+'> / dsl::digit<>) >> dsl::p<integer>;
    return ( hex_integer | regular_integer ) + dsl::eof;
  }();

  static constexpr auto value = lexy::forward<int64_t>;
};
```

**Reply by foonathan** [reply]
> but I am not quite sure why the dsl::peek is required.

The choice operator requires a branch condition to determine whether or not it should attempt to parse `hex_integer` or `regular_integer`. In the case of `hex_integer` that is the existence of the `0x` prefix, but since `dsl::sign` can match `+`, `-` or nothing it can't be used as a branch condition (i.e. the left hand side of `operator>>`). As such, you need to manually specify which character to peek. In this particular case you can also use `dsl::else_` instead which is a catch-all condition that always matches. More info here: https://lexy.foonathan.net/learn/branching/

Also note that the nested production is not necessary. You can just use `dsl::sign + dsl::integer<int64_t>` inline and use `lexy::as_integer` as the callback.
